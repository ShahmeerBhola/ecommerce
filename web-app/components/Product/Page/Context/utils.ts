import isEmpty from 'lodash/isEmpty';

import type { TProductContext } from './types';

export const getSingleAttributes = (_attributes?: StringMap<string[]>): StringMap => {
    let singleAttributes = {};

    for (const [key, value] of Object.entries(_attributes || {})) {
        if (value.length === 1) {
            singleAttributes = {
                ...singleAttributes,
                [key]: value[0],
            };
        }
    }

    return singleAttributes;
};

export const getAvailableAttributes = (
    variations: WordPress.ProductVariation[],
    newAttributes: StringMap,
): StringMap<string[]> => {
    const availableAttribute: StringMap<string[]> = {};

    for (const { attributes } of variations) {
        let isAvailableAttribute = true;

        for (const [key, value] of Object.entries(newAttributes)) {
            isAvailableAttribute = isAvailableAttribute && attributes[key] === value;
        }

        if (isAvailableAttribute) {
            for (const [key, value] of Object.entries(attributes)) {
                const currentAttribute = availableAttribute[key] || [];

                if (currentAttribute.indexOf(value) > -1) {
                    continue;
                }

                availableAttribute[key] = [...currentAttribute, value];
            }
        }
    }

    return availableAttribute;
};

const getAttributes = (
    variations: WordPress.ProductVariation[],
    selectedAttributes: {} | StringMap,
): TProductContext.CheckAttributesType => {
    const availableAttributes = getAvailableAttributes(variations, selectedAttributes);
    const singleAttributes = getSingleAttributes(availableAttributes);

    return {
        currentAttributes: {
            ...selectedAttributes,
            ...singleAttributes,
        },
        availableAttributes,
    };
};

export const checkAttributes = (state: TProductContext.State, payload: any): TProductContext.CheckAttributesType => {
    const { currentAttributes, product } = state;
    const { variations } = product as WordPress.Product;
    const newAttributes = {
        ...currentAttributes,
        ...payload,
    };

    const attributes = getAttributes(variations, newAttributes);

    if (isEmpty(attributes.availableAttributes)) {
        const resetAttributes = {
            ...payload,
        };

        return getAttributes(variations, resetAttributes);
    }

    return attributes;
};

export const isActiveAttribute = (attributes: StringMap, currentAttributes: StringMap): boolean => {
    return Object.entries(attributes).reduce<boolean>((isSelectedAttribute, attribute) => {
        const [key, value] = attribute;
        const matchedAttribute = value === '' || currentAttributes[key] === value;

        return isSelectedAttribute && matchedAttribute;
    }, true);
};
