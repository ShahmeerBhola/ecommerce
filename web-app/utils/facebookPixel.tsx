import React from 'react';
import config from '../config';

const { facebookPixelID } = config;

// https://www.facebook.com/business/help/402791146561655?id=1205376682832142
const event = <P,>(eventName: string, params?: P): void => {
  if (facebookPixelID !== null) {
    if (params !== undefined) {
      window.fbq('track', eventName, params);
    } else {
      window.fbq('track', eventName);
    }
  }
};

export const addLineItemToCart = ({ sku, name, price }: WordPress.ProductBase, type = 'line_item'): void =>
  event<facebook.Pixel.AddToCartParameters>('AddToCart', {
    content_name: name,
    content_type: type,
    content_ids: [sku],
    value: price,
    currency: config.defaultCurrencyCode,
  });

export const addWheelTirePackageToCart = ({ wheel, tire, addOns }: Cart.WheelTirePackage): void => {
  const addWheelTirePackageItemToCart = (product: WordPress.ProductBase): void =>
    addLineItemToCart(product, 'wheel_tire_package');

  addWheelTirePackageItemToCart(wheel);
  addWheelTirePackageItemToCart(tire);
  addOns.map((addOn) => addWheelTirePackageItemToCart(addOn));
};

export const initiateCheckout = (): void => event('InitiateCheckout');
export const completePurchase = (params: facebook.Pixel.PurchaseParameters): void =>
  event<facebook.Pixel.PurchaseParameters>('Purchase', params);

export const FacebookPixelScripts: React.FunctionComponent = () => {
  {
    if (facebookPixelID !== null) {
      return (
        <>
          <script
            dangerouslySetInnerHTML={{
              __html: `
            !function(f,b,e,v,n,t,s)
{if(f.fbq)return;n=f.fbq=function(){n.callMethod ?
              n.callMethod.apply(n, arguments) : n.queue.push(arguments)};
if(!f._fbq)f._fbq=n;n.push=n;n.loaded=!0;n.version='2.0';
n.queue=[];t=b.createElement(e);t.async=!0;
t.src=v;s=b.getElementsByTagName(e)[0];
s.parentNode.insertBefore(t,s)}(window,document,'script',
'https://connect.facebook.net/en_US/fbevents.js');
 fbq('init', '${facebookPixelID}');
fbq('track', 'PageView');
          `,
            }}
          />
          <noscript>
            <img
              height="1"
              width="1"
              src={`https://www.facebook.com/tr?id=${facebookPixelID}&ev=PageView
&noscript=1`}
            />
          </noscript>
        </>
      );
    }
  }
  return null;
};
