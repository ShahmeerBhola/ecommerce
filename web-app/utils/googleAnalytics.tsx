import React from 'react';
import config from '../config';
import { isBrowserSide } from './';

const { googleAnalyticsTrackingID, googleAdsConversionID } = config;

// https://developers.google.com/analytics/devguides/collection/gtagjs/pages
// https://github.com/vercel/next.js/tree/canary/examples/with-google-analytics
export const pageview = (url: string): void => {
  if (googleAnalyticsTrackingID !== null && isBrowserSide()) {
    window.gtag('config', googleAnalyticsTrackingID, {
      page_path: url,
    });
  }
};

// https://developers.google.com/analytics/devguides/collection/gtagjs/events
const event = (action: string, eventParams: any): void => {
  if (googleAnalyticsTrackingID !== null && isBrowserSide()) {
    window.gtag('event', action, eventParams);
  }
};

const conversionEvent = (sendTo: string, eventParams?: StringMap): void =>
  event('conversion', { send_to: sendTo, ...eventParams });

export const sendOrderConfirmationConversionEvent = (orderId: string): void =>
  conversionEvent('AW-479563831/zZZcCPqr3-gBELeg1uQB', { transaction_id: orderId });

export const sendContactFormConversionEvent = (): void => conversionEvent('AW-479563831/6i-DCKyL6-gBELeg1uQB');

export const sendRequestForQuoteEvent = (): void => conversionEvent('AW-479563831/C9PsCJv7_-gBELeg1uQB');

// https://developers.google.com/analytics/devguides/collection/gtagjs/enhanced-ecommerce
export const sendEnhancedEcommerceEvent = (
  eee: GoogleAnalytics.EnhancedEcommerceEvent,
  items: GoogleAnalytics.Product[],
): void => event(eee, { items });

export const sendEnhancedEcommercePurchaseEvent = (data: GoogleAnalytics.PurchaseEvent): void =>
  event('purchase', data);

export const sendEnhancedEcommerceCheckoutOptionEvent = (data: GoogleAnalytics.CheckoutProgressEvent): void =>
  event('set_checkout_option', data);

export const enhancedEcommerceLineItemListName = 'Line Items';
export const enhancedEcommerceWheelTirePackageListName = 'Wheel Tire Package';
export const enhancedEcommerceWheelTirePackageAddOnsListName = 'Wheel Tire Package Add Ons';

export const GoogleAnalyticsAndAdsScripts: React.FunctionComponent = () => {
  {
    if (googleAnalyticsTrackingID !== null && googleAdsConversionID !== null) {
      return (
        <>
          <script async src={`https://www.googletagmanager.com/gtag/js?id=${googleAnalyticsTrackingID}`} />
          <script
            dangerouslySetInnerHTML={{
              __html: `
            window.dataLayer = window.dataLayer || [];
            function gtag(){dataLayer.push(arguments);}
            gtag('js', new Date());
            gtag('config', '${googleAnalyticsTrackingID}', {
              page_path: window.location.pathname,
            });
            gtag('config', '${googleAdsConversionID}')
          `,
            }}
          />
        </>
      );
    }
  }
  return null;
};
