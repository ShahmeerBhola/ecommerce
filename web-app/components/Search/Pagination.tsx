import React, { FC } from 'react';
import { useRouter } from 'next/router';
import Text from '../Text';
import { RightCarrotIcon, LeftCarrotIcon } from '../Icons';
import { modifyQueryParam, calculatePageNumbersToShow } from '../../utils';
import config from '../../config';

type Props = {
    pagination: WordPress.Pagination;
};

const Pagination: FC<Props> = ({ pagination: { total_pages, page } }) => {
    const router = useRouter();
    const { paginationPagesToShow } = config;
    const numbersToShow = calculatePageNumbersToShow(total_pages, page);

    const changePage = (page: number): void => modifyQueryParam(router, 'page', page.toString());

    if (numbersToShow.length === 0) {
        return null;
    }

    return (
        <div className="inline-flex items-center float-right">
            <Text className="mr-5" size="text-sm">
                Page
            </Text>
            {page !== 1 && total_pages > paginationPagesToShow ? (
                <div className="mx-3">
                    <LeftCarrotIcon onClick={(): void => changePage(page - 1)} />
                </div>
            ) : null}
            {numbersToShow.map((number) => {
                const isActive = number === page;
                return (
                    <div
                        className={`mx-1 py-1 px-3 cursor-pointer${
                            isActive ? ' bg-white border border-gray-80 border-solid' : ''
                        }`}
                        key={number}
                        onClick={(): void => changePage(number)}
                    >
                        <Text size="text-sm" color={isActive ? 'text-red-100' : 'text-gray-100'}>
                            {number}
                        </Text>
                    </div>
                );
            })}
            {page !== total_pages && total_pages > paginationPagesToShow ? (
                <div className="mx-3">
                    <RightCarrotIcon onClick={(): void => changePage(page + 1)} />
                </div>
            ) : null}
        </div>
    );
};

export default Pagination;
