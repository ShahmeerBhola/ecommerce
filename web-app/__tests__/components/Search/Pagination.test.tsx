import React from 'react';
import renderer from 'react-test-renderer';
import { generateMockPagination } from '../../utils';
import Pagination from '../../../components/Search/Pagination';

describe('Pagination component', () => {
  test('it renders properly when there is lots of pages', () => {
    const tree = renderer.create(<Pagination pagination={generateMockPagination(5005, 201, 1, 25)} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('it renders properly when there is only one page', () => {
    const tree = renderer.create(<Pagination pagination={generateMockPagination(3, 1, 1, 25)} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('it renders properly when there are no pages', () => {
    const tree = renderer.create(<Pagination pagination={generateMockPagination(0, 0, 1, 25)} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('it renders properly when we are on page 3', () => {
    const tree = renderer.create(<Pagination pagination={generateMockPagination(5005, 201, 3, 25)} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('it renders properly when we are on the second to last page', () => {
    const tree = renderer.create(<Pagination pagination={generateMockPagination(5005, 201, 200, 25)} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('it renders properly when we are on the last page', () => {
    const tree = renderer.create(<Pagination pagination={generateMockPagination(5005, 201, 201, 25)} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
