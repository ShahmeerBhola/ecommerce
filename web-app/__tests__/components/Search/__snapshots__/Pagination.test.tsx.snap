// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination component it renders properly when there are no pages 1`] = `null`;

exports[`Pagination component it renders properly when there is lots of pages 1`] = `
<div
  className="inline-flex items-center float-right"
>
  <p
    className="font-secondary font-normal text-gray-100 text-base mr-5"
  >
    Page
  </p>
  <div
    className="mx-2 py-2 px-3 cursor-pointer bg-white border border-gray-80 border-solid"
    onClick={[Function]}
  >
    <p
      className="font-secondary font-normal text-red-100 text-base"
    >
      1
    </p>
  </div>
  <div
    className="mx-2 py-2 px-3 cursor-pointer"
    onClick={[Function]}
  >
    <p
      className="font-secondary font-normal text-gray-100 text-base"
    >
      2
    </p>
  </div>
  <div
    className="mx-2 py-2 px-3 cursor-pointer"
    onClick={[Function]}
  >
    <p
      className="font-secondary font-normal text-gray-100 text-base"
    >
      3
    </p>
  </div>
  <div
    className="mx-2 py-2 px-3 cursor-pointer"
    onClick={[Function]}
  >
    <p
      className="font-secondary font-normal text-gray-100 text-base"
    >
      4
    </p>
  </div>
  <div
    className="mx-2 py-2 px-3 cursor-pointer"
    onClick={[Function]}
  >
    <p
      className="font-secondary font-normal text-gray-100 text-base"
    >
      5
    </p>
  </div>
  <div
    className="mx-3"
  >
    <svg
      className="cursor-pointer"
      fill="none"
      height="31"
      onClick={[Function]}
      viewBox="0 0 19 31"
      width="19"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M18 15.5L2.84783 30L1 28.2317L14.3043 15.5L1 2.76829L2.84783 1L18 15.5Z"
        fill="#F23939"
        stroke="#F23939"
      />
    </svg>
  </div>
</div>
`;

exports[`Pagination component it renders properly when there is only one page 1`] = `
<div
  className="inline-flex items-center float-right"
>
  <p
    className="font-secondary font-normal text-gray-100 text-base mr-5"
  >
    Page
  </p>
  <div
    className="mx-2 py-2 px-3 cursor-pointer bg-white border border-gray-80 border-solid"
    onClick={[Function]}
  >
    <p
      className="font-secondary font-normal text-red-100 text-base"
    >
      1
    </p>
  </div>
</div>
`;

exports[`Pagination component it renders properly when we are on page 3 1`] = `
<div
  className="inline-flex items-center float-right"
>
  <p
    className="font-secondary font-normal text-gray-100 text-base mr-5"
  >
    Page
  </p>
  <div
    className="mx-3"
  >
    <svg
      className="cursor-pointer"
      fill="none"
      height="31"
      onClick={[Function]}
      viewBox="0 0 19 31"
      width="19"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M1 15.5L16.1522 30L18 28.2317L4.69565 15.5L18 2.76829L16.1522 1L1 15.5Z"
        fill="#F23939"
        stroke="#F23939"
      />
    </svg>
  </div>
  <div
    className="mx-2 py-2 px-3 cursor-pointer"
    onClick={[Function]}
  >
    <p
      className="font-secondary font-normal text-gray-100 text-base"
    >
      1
    </p>
  </div>
  <div
    className="mx-2 py-2 px-3 cursor-pointer"
    onClick={[Function]}
  >
    <p
      className="font-secondary font-normal text-gray-100 text-base"
    >
      2
    </p>
  </div>
  <div
    className="mx-2 py-2 px-3 cursor-pointer bg-white border border-gray-80 border-solid"
    onClick={[Function]}
  >
    <p
      className="font-secondary font-normal text-red-100 text-base"
    >
      3
    </p>
  </div>
  <div
    className="mx-2 py-2 px-3 cursor-pointer"
    onClick={[Function]}
  >
    <p
      className="font-secondary font-normal text-gray-100 text-base"
    >
      4
    </p>
  </div>
  <div
    className="mx-2 py-2 px-3 cursor-pointer"
    onClick={[Function]}
  >
    <p
      className="font-secondary font-normal text-gray-100 text-base"
    >
      5
    </p>
  </div>
  <div
    className="mx-3"
  >
    <svg
      className="cursor-pointer"
      fill="none"
      height="31"
      onClick={[Function]}
      viewBox="0 0 19 31"
      width="19"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M18 15.5L2.84783 30L1 28.2317L14.3043 15.5L1 2.76829L2.84783 1L18 15.5Z"
        fill="#F23939"
        stroke="#F23939"
      />
    </svg>
  </div>
</div>
`;

exports[`Pagination component it renders properly when we are on the last page 1`] = `
<div
  className="inline-flex items-center float-right"
>
  <p
    className="font-secondary font-normal text-gray-100 text-base mr-5"
  >
    Page
  </p>
  <div
    className="mx-3"
  >
    <svg
      className="cursor-pointer"
      fill="none"
      height="31"
      onClick={[Function]}
      viewBox="0 0 19 31"
      width="19"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M1 15.5L16.1522 30L18 28.2317L4.69565 15.5L18 2.76829L16.1522 1L1 15.5Z"
        fill="#F23939"
        stroke="#F23939"
      />
    </svg>
  </div>
  <div
    className="mx-2 py-2 px-3 cursor-pointer"
    onClick={[Function]}
  >
    <p
      className="font-secondary font-normal text-gray-100 text-base"
    >
      199
    </p>
  </div>
  <div
    className="mx-2 py-2 px-3 cursor-pointer"
    onClick={[Function]}
  >
    <p
      className="font-secondary font-normal text-gray-100 text-base"
    >
      200
    </p>
  </div>
  <div
    className="mx-2 py-2 px-3 cursor-pointer bg-white border border-gray-80 border-solid"
    onClick={[Function]}
  >
    <p
      className="font-secondary font-normal text-red-100 text-base"
    >
      201
    </p>
  </div>
</div>
`;

exports[`Pagination component it renders properly when we are on the second to last page 1`] = `
<div
  className="inline-flex items-center float-right"
>
  <p
    className="font-secondary font-normal text-gray-100 text-base mr-5"
  >
    Page
  </p>
  <div
    className="mx-3"
  >
    <svg
      className="cursor-pointer"
      fill="none"
      height="31"
      onClick={[Function]}
      viewBox="0 0 19 31"
      width="19"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M1 15.5L16.1522 30L18 28.2317L4.69565 15.5L18 2.76829L16.1522 1L1 15.5Z"
        fill="#F23939"
        stroke="#F23939"
      />
    </svg>
  </div>
  <div
    className="mx-2 py-2 px-3 cursor-pointer"
    onClick={[Function]}
  >
    <p
      className="font-secondary font-normal text-gray-100 text-base"
    >
      198
    </p>
  </div>
  <div
    className="mx-2 py-2 px-3 cursor-pointer"
    onClick={[Function]}
  >
    <p
      className="font-secondary font-normal text-gray-100 text-base"
    >
      199
    </p>
  </div>
  <div
    className="mx-2 py-2 px-3 cursor-pointer bg-white border border-gray-80 border-solid"
    onClick={[Function]}
  >
    <p
      className="font-secondary font-normal text-red-100 text-base"
    >
      200
    </p>
  </div>
  <div
    className="mx-2 py-2 px-3 cursor-pointer"
    onClick={[Function]}
  >
    <p
      className="font-secondary font-normal text-gray-100 text-base"
    >
      201
    </p>
  </div>
  <div
    className="mx-3"
  >
    <svg
      className="cursor-pointer"
      fill="none"
      height="31"
      onClick={[Function]}
      viewBox="0 0 19 31"
      width="19"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M18 15.5L2.84783 30L1 28.2317L14.3043 15.5L1 2.76829L2.84783 1L18 15.5Z"
        fill="#F23939"
        stroke="#F23939"
      />
    </svg>
  </div>
</div>
`;
